### PR TITLE
Implementation of ZCL AddNetwork and EnableNetwork methods

### LINKED ISSUE #15: RPC Support for ZCL Network Commissioning
#### Problem
Need ZCL Network Commissioning Commands

 #### Proposed Solution
Expose ZCL Network Commission and Network Enable commands over RPC

### PR DESCRIPTION
**Pull request by nirav1911**
Problem
Implemented ZCL AddNetwork and EnableNetwork methods

Summary of Changes
Implemented ZCL AddNetwork and EnableNetwork methods
 with RPC

Fixed #15

### CODE REVIEW

File: src/controller/python/chip-device-ctrl.py [hunk 1]
```diff
@@ -629,17 +624,73 @@ def emptyline(self):
...
+
+def resolve(fabric_id: int, node_id: int) -> Dict[str, Any]:
     try:
         __check_supported_os()
-        device_manager.do_blescan("")
+        err = device_manager.devCtrl.ResolveNode(fabric_id, node_id)
+        if err == 0:
+            address = device_manager.devCtrl.GetAddressAndPort(int(args[1]))
+            address = "{}:{}".format(
+                *address) if address else "unknown"
```
Comment (cjandhyala) [inline src/controller/python/chip-device-ctrl.py:640]: if address "unknown" don't we need to send back  __get_response_dict(status = StatusCodeEnum.FAILED, error = str(ex)) ?

File: src/controller/python/chip-device-ctrl.py [hunk 2]
```diff
@@ -43,7 +43,6 @@
 from xmlrpc.server import SimpleXMLRPCServer
 from enum import Enum
 from typing import Any, Dict,Optional
-
 # Extend sys.path with one or more directories, relative to the location of the
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:46]: This line removal looks like its outside our zone of operation?

File: src/controller/python/chip-device-ctrl.py [hunk 3]
```diff
@@ -141,8 +137,7 @@ def FormatZCLArguments(args, command):
 
 class DeviceMgrCmd(Cmd):
     def __init__(self, rendezvousAddr=None, controllerNodeId=0, bluetoothAdapter=None):
-        self.lastNetworkId = None
-
+        self.lastNetworkId = None
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:140]: Same as above, we need to be careful with line removals due to potential upstream issues

File: src/controller/python/chip-device-ctrl.py [hunk 4]
```diff
@@ -629,17 +624,73 @@ def emptyline(self):
...
+        if err == 0:
+            address = device_manager.devCtrl.GetAddressAndPort(int(args[1]))
+            address = "{}:{}".format(
+                *address) if address else "unknown"
+            
+            return __get_response_dict(status = StatusCodeEnum.SUCCESS, result = {'address': address})
+    except exceptions.ChipStackException as ex:
+        return __get_response_dict(status = StatusCodeEnum.FAILED, error = str(ex))
+    except Exception as e:
+        return __get_response_dict(status = StatusCodeEnum.FAILED, error = str(e))
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:646]: seems like `except Exception as e:` will be able to catch the `exceptions.ChipStackException` as well

File: src/controller/python/chip-device-ctrl.py [hunk 5]
```diff
@@ -629,17 +624,73 @@ def emptyline(self):
...
+        if err != 0:
+            return __get_response_dict(status = StatusCodeEnum.FAILED)
+        elif res != None:
+            return __get_response_dict(status = StatusCodeEnum.SUCCESS, result = str(res))
+        else:
+            return __get_response_dict(status = StatusCodeEnum.SUCCESS)
         
+    except exceptions.ChipStackException as ex:
+        return __get_response_dict(status = StatusCodeEnum.FAILED, error = str(ex))
+    except Exception as e:
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:666]: Same comment as above

File: src/controller/python/chip-device-ctrl.py [hunk 6]
```diff
@@ -629,17 +624,73 @@ def emptyline(self):
...
+        args['breadcrumb'] = breadcrumb
+        args['timeoutMs'] = timeoutMs 
+  
+        err, res = device_manager.devCtrl.ZCLSend("NetworkCommissioning", "EnableNetwork", node_id, endpoint_id, group_id, args, blocking=True)
+        if err != 0:
+            return __get_response_dict(status = StatusCodeEnum.FAILED)
+        elif res != None:
+            return __get_response_dict(status = StatusCodeEnum.SUCCESS, result = str(res))
+        else:
+            return __get_response_dict(status = StatusCodeEnum.SUCCESS)
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:683]: Does python allow setting a method parameter like `result` here to None?
Seems like if result is optional these two elif and else conditions can be combined

File: src/controller/python/chip-device-ctrl.py [hunk 7]
```diff
@@ -674,23 +725,27 @@ def ble_connect(discriminator: int, pin_code: int, node_id: int) -> Dict[str, an
...
         return __get_response_dict(status = StatusCodeEnum.FAILED, error = str(e))
 
-def ip_connect(ip_address: string, pin_code: int, node_id: int) -> Dict[str, any]:
+def ip_connect(ip_address: string, pin_code: int, node_id: int) -> Dict[str, Any]:
     try:
         __check_supported_os()
         device_manager.devCtrl.ConnectIP(ip_address.encode("utf-8"), pin_code, node_id)
         return __get_response_dict(status = StatusCodeEnum.SUCCESS)
     except exceptions.ChipStackException as ex:
+        print(str(ex))
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:734]: Debugging print?

File: src/controller/python/chip-device-ctrl.py (hunk 2 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Resolved

File: src/controller/python/chip-device-ctrl.py (hunk 1 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Resolved

File: src/controller/python/chip-device-ctrl.py (hunk 4 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py:653]: Resolved

File: src/controller/python/chip-device-ctrl.py (hunk 5 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py:671]: Resolved

File: src/controller/python/chip-device-ctrl.py (hunk 6 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Resolved

File: src/controller/python/chip-device-ctrl.py (hunk 7 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Removed

File: src/controller/python/chip-device-ctrl.py [hunk 8]
```diff
@@ -711,7 +762,7 @@ def __check_supported_os()-> bool:
     raise Exception("OS Not Supported")
 
 ######--------------------------------------------------######
-
+
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:765]: Unnecessary modification outside the ###----### scope

**Review by Xiole-KKE — APPROVED** [review]
LGTM

File: src/controller/python/chip-device-ctrl.py [hunk 9]
```diff
@@ -635,11 +635,63 @@ def echo_alive(message):
...
+            return __get_response_dict(status = StatusCodeEnum.FAILED)
+        elif res != None:
+            return __get_response_dict(status = StatusCodeEnum.SUCCESS, result = str(res))
+        else:
+            return __get_response_dict(status = StatusCodeEnum.SUCCESS)
+
+    except Exception as e:
+        return __get_response_dict(status = StatusCodeEnum.FAILED, error = str(e))
+
+def zcl_enable_network(node_id: int, ssid:str, endpoint_id: Optional[int] = 1, group_id: Optional[int] = 0, breadcrumb: Optional[int] = 0, timeoutMs: Optional[int] = 1000) -> Dict[str, Any]:
```
Comment (mikaelhm) [inline src/controller/python/chip-device-ctrl.py:674]: Would it be easier to expose the `ZCLSend` api directly, so we don't need to add all the APIs in the RPC, but define them in our test harness?

I think that would be a bit easier to maintain, but maybe I am missing something.

**Review by mikaelhm — CHANGES_REQUESTED** [review]

File: src/controller/python/chip-device-ctrl.py (hunk 9 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py:674]: This was discussed with @wahajsyed . We can certainly add zclSend in future. Currently we have added basic commands and see in next iteration if `zclSend` makes sense or not.

File: src/controller/python/chip-device-ctrl.py [hunk 10]
```diff
@@ -635,11 +635,63 @@ def echo_alive(message):
...
+        args['credentials'] = password.encode("utf-8") + b'\x00'
+        args['breadcrumb'] = breadcrumb
+        args['timeoutMs'] = timeoutMs 
+        err, res = device_manager.devCtrl.ZCLSend("NetworkCommissioning", "AddWiFiNetwork", node_id, endpoint_id, group_id, args, blocking=True)
+        if err != 0:
+            return __get_response_dict(status = StatusCodeEnum.FAILED)
+        elif res != None:
+            return __get_response_dict(status = StatusCodeEnum.SUCCESS, result = str(res))
+        else:
+            return __get_response_dict(status = StatusCodeEnum.SUCCESS)
```
[the remark this answers is not in the document]
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py:669]: based on the current `do_zcl` code, if res is `None`, it shows as success. I'm not able to test zcl and resolve commands so we might need to change the code based on the output.

File: src/controller/python/chip-device-ctrl.py [hunk 11]
```diff
@@ -635,11 +635,63 @@ def echo_alive(message):
...
 
-def ble_scan() -> Dict[Any, Any]:
+
+def resolve(fabric_id: int, node_id: int) -> Dict[str, Any]:
     try:
         __check_supported_os()
-        device_manager.do_blescan("")
+        err = device_manager.devCtrl.ResolveNode(fabric_id, node_id)
+        if err == 0:
+            address = device_manager.devCtrl.GetAddressAndPort(int(args[1]))
```
[the remark this answers is not in the document]
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py:644]: updated.

File: src/controller/python/chip-device-ctrl.py [hunk 12]
```diff
@@ -699,7 +750,7 @@ def start_rpc_server():
             print("\nKeyboard interrupt received, exiting.")
             sys.exit(0)
 
-def __get_response_dict(status: StatusCodeEnum, result: Optional[Dict[Any, Any]] = None, error:Optional[str] = None) -> Dict [str, Any]:
+def __get_response_dict(status: StatusCodeEnum, result: Optional[Dict[Any, Any]] = "", error:Optional[str] = "") -> Dict [str, Any]:
```
Comment (mikaelhm) [inline src/controller/python/chip-device-ctrl.py:753]: I think we should consider leaving the None in here; and set the allow_none=true on the XMLRPCserver. 

I had the same issue when testing our PASE RPC. 

But I think the API is cleaner if it returns None.

**Review by mikaelhm — APPROVED** [review]
I just left one comment. But otherwise looks good now.

File: src/controller/python/chip-device-ctrl.py (hunk 12 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Added allow_none=true on RPC Server.

**Pull request merged**